Reimplement invaliding `Notification` attributes cache

crazytonyli opened the pull request:
The original implementation overrides `NSManagedObject.willChangeValue` which must not be overridden as mentioned in [the API doc](https://developer.apple.com/documentation/coredata/nsmanagedobject/1506229-willchangevalue).

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
None.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

crazytonyli committed:
Reimplement invaliding `Notification` attributes cache

The original implementation overrides `NSManagedObject.willChangeValue`
which must not be overridden as mentioned in the API doc.

See https://developer.apple.com/documentation/coredata/nsmanagedobject/1506229-willchangevalue

crazytonyli committed:
Fix a linting issue

crazytonyli commented on WordPress/Classes/Models/Notifications/Notification.swift:
```diff
@@ -412,3 +411,21 @@ extension Notification: Notifiable {
+        //
+        guard notification.managedObjectContext?.concurrencyType == .mainQueueConcurrencyType else {
+            return
+        }
```
The original observer logic is kept, including this one here. But I think not invaliding the cache just because the notification object is in a background context is problematic, see [the unit test  below](https://github.com/wordpress-mobile/WordPress-iOS/pull/19395/files#diff-b159138f3547f7e8daa93740188dd5d1432f1314e123260fefedfa94ae8d99abR254) for more.

crazytonyli committed:
Query the notification from DB instead of use the created instance

mokagio commented on WordPress/Classes/Models/Notifications/Notification.swift:
```diff
@@ -81,10 +81,29 @@ class Notification: NSManagedObject {
     ///
     fileprivate var cachedHeaderAndBodyContentGroups: [FormattableContentGroup]?
 
+    private let cachedAttributesObserver: NotificationCachedAttributesObserver? = nil
```
Is the explicit `= nil` assignment intentional?

I think the compiler is already doing that for us.

crazytonyli replied:
Not for `let` variable apparently. But this property is changed to a `var` and I've removed the `nil` assignment.

mokagio replied:
> Not for `let` variable apparently

Blimey! I didn't notice it was a `let` 😳

mokagio commented on WordPress/Classes/Models/Notifications/Notification.swift:
```diff
@@ -81,10 +81,29 @@ class Notification: NSManagedObject {
+        cachedAttributesObserver = NotificationCachedAttributesObserver()
+        for attr in Notification.cachedAttributes {
+            addObserver(cachedAttributesObserver!, forKeyPath: attr, context: nil)
+        }
```
What do you think of doing the following, to remove the force unwrap?

```suggestion
        let cachedAttributesObserver = NotificationCachedAttributesObserver()
        for attr in Notification.cachedAttributes {
            addObserver(cachedAttributesObserver, forKeyPath: attr, context: nil)
        }
        self.cachedAttributesObserver = cachedAttributesObserver
```

The force unwrap is of course okay in the way it's used now, but who's to tell how the code will evolve? There might end up being a code path in which it's `nil` in the future.

Maybe I'm overthinking it 😅

crazytonyli replied:
👍 addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/19395/commits/da753bdff58fa9e39efd00c9e758efc50fc308ef. It's always good to not use force unwrap.

mokagio reviewed (approved):
Love how you wrote a test to show the threading issue!

I started looking at the git blame, then got interrupted, then side tracked, and I'm only now coming back to my WIP-review.

I didn't get to the bottom of how that `.mainQueueConcurrencyType` came about, but the comment states it was a matter of optimization. Maybe it was a case of _premature_ optimization?

I'd be okay with removing the restriction and see how the app behaves.

crazytonyli committed:
Register KVO observer in `awakeFromFetch`

This will reduce amount of KVO observer calls, now that only the fetched
instances have KVO observer registered, not the inserted ones.

crazytonyli committed:
Keep the original behaviour - resetting cache before changes

crazytonyli committed:
Minor clean up
